File: Annual_Police_Data.py
# -*- coding: utf-8 -*-
"""
Created on Sun Nov 26 19:28:52 2023

@author: natal
"""

import os
import pandas as pd
import matplotlib.pyplot as plt


#base_path = r'C:\Users\natal\Desktop\Smriti\SARC_research'
base_path = r'C:\Users\natal\Desktop\GitHub\SARC_research'

main_folder = os.path.join(base_path, 'Police_Data')

police_df_2011 = pd.DataFrame(columns=['Month', 'Falls within', 'Crime type'])

# Define the root directory where your folders are located
root_directory = main_folder

# Iterate through all items in the root directory
for item in os.listdir(root_directory):
    # Check if the item is a directory and contains '2011'
    if os.path.isdir(os.path.join(root_directory, item)) and '2011' in item:
        # Iterate through each file in the directory
        for file in os.listdir(os.path.join(root_directory, item)):
            # Check if the file ends with '-street.csv'
            if file.endswith('-street.csv'):
                file_path = os.path.join(root_directory, item, file)
                # Read only the specified columns from the CSV file
                df = pd.read_csv(file_path, usecols=['Month', 'Falls within', 'Crime type'])
                # Append it to the police_df_2011 DataFrame
                police_df_2011 = police_df_2011.append(df, ignore_index=True)


### Get summary stats for crime type in 2011

## Unique Responses
# Get unique responses in the "Crime type" column
unique_crime_types_2011 = police_df_2011['Crime type'].unique()

# Print the unique crime types
print(unique_crime_types_2011)

# Get unique responses in the "Falls within" column
unique_police_neighborhoods_2011 = police_df_2011['Falls within'].unique()

# Print the unique crime types
print(unique_police_neighborhoods_2011)

# %%

def create_police_df_for_year(year, root_directory):

    # Initialize the DataFrame with specified columns
    police_df = pd.DataFrame(columns=['Month', 'Falls within', 'Crime type'])

    # Iterate through all items in the root directory
    for item in os.listdir(root_directory):
        # Check if the item is a directory and contains the specified year
        if os.path.isdir(os.path.join(root_directory, item)) and str(year) in item:
            # Iterate through each file in the directory
            for file in os.listdir(os.path.join(root_directory, item)):
                # Check if the file ends with '-street.csv'
                if file.endswith('-street.csv'):
                    file_path = os.path.join(root_directory, item, file)
                    # Read only the specified columns from the CSV file
                    df = pd.read_csv(file_path, usecols=['Month', 'Falls within', 'Crime type'])
                    # Append it to the police_df DataFrame
                    police_df = police_df.append(df, ignore_index=True)

    return police_df

police_df_2012 = create_police_df_for_year(2012, root_directory)
police_df_2013 = create_police_df_for_year(2013, root_directory)
police_df_2014 = create_police_df_for_year(2014, root_directory)
police_df_2015 = create_police_df_for_year(2015, root_directory)
police_df_2016 = create_police_df_for_year(2016, root_directory)
police_df_2017 = create_police_df_for_year(2017, root_directory)
police_df_2018 = create_police_df_for_year(2018, root_directory)
police_df_2019 = create_police_df_for_year(2019, root_directory)
police_df_2020 = create_police_df_for_year(2020, root_directory)
police_df_2021 = create_police_df_for_year(2021, root_directory)
police_df_2022 = create_police_df_for_year(2022, root_directory)

# %%

def print_unique_values(year, dataframes):
    """
    Print unique values in specific columns for the given year.

    Parameters:
    year (int): The year for which to print unique values.
    dataframes (dict): A dictionary of DataFrames, keyed by year.
    """
    # Access the DataFrame for the specified year
    df = dataframes.get(year)
    if df is not None:
        # Get and print unique values for "Crime type"
        unique_crime_types = df['Crime type'].unique()
        print(f"Unique Crime Types for {year}:")
        print(unique_crime_types)

        # Get and print unique values for "Falls within"
        unique_police_neighborhoods = df['Falls within'].unique()
        print(f"\nUnique Police Neighborhoods for {year}:")
        print(unique_police_neighborhoods)
    else:
        print(f"No data available for the year {year}")



dataframes = {
    2011: police_df_2011,
    2012: police_df_2012,
    2013: police_df_2013,
    2014: police_df_2014,
    2015: police_df_2015,
    2016: police_df_2016,
    2017: police_df_2017,
    2018: police_df_2018,
    2019: police_df_2019,
    2020: police_df_2020,
    2021: police_df_2021,
    2022: police_df_2022,
}

print_unique_values(2012, dataframes)
print_unique_values(2013, dataframes)
print_unique_values(2014, dataframes)
print_unique_values(2015, dataframes)
print_unique_values(2016, dataframes)
print_unique_values(2017, dataframes)
print_unique_values(2018, dataframes)
print_unique_values(2019, dataframes)
print_unique_values(2020, dataframes)
print_unique_values(2021, dataframes)
print_unique_values(2022, dataframes)

# %%
# Initialized a DataFrame to store the counts for each year
monthly_counts_df = pd.DataFrame(columns=['Year', 'January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'])

# Function to process a year's data
def process_year(df, year):
    # Filter for 'Violence and sexual offences'
    filtered_df = df[df['Crime type'] == 'Violence and sexual offences']
    
    # Count occurrences for each month
    monthly_counts = filtered_df['Month'].value_counts().sort_index()
    
    # Create a new row with year and monthly counts
    new_row = {'Year': year}
    for month in monthly_counts.index:
        # Assuming the 'Month' column is in 'YYYY-MM' format
        month_name = pd.to_datetime(month).strftime('%B')
        new_row[month_name] = monthly_counts[month]
    
    return new_row


for year in range(2013, 2023):  # Adjust range as needed
    yearly_df = globals()[f'police_df_{year}']
    monthly_counts_df = monthly_counts_df.append(process_year(yearly_df, year), ignore_index=True)
    
# %%

def calculate_annual_statistics_from_monthly_counts(monthly_counts_df):
    """
    Calculate the annual min, max, mean, and std deviation for each year in the monthly_counts_df.

    Parameters:
    monthly_counts_df (pd.DataFrame): A DataFrame with monthly crime counts, indexed by year.

    Returns:
    pd.DataFrame: A DataFrame containing the statistics for each year.
    """
    # Initialize a DataFrame to store the statistics
    stats_df = pd.DataFrame(columns=['Year', 'Min', 'Max', 'Mean', 'StdDev'])

    # Iterate over each row in the monthly_counts_df
    for index, row in monthly_counts_df.iterrows():
        year = row['Year']
        monthly_counts = row.drop('Year')  # Exclude the Year column

        min_value = monthly_counts.min()
        max_value = monthly_counts.max()
        mean_value = monthly_counts.mean()
        std_dev = monthly_counts.std()

        # Append the statistics for this year to the DataFrame
        stats_df = stats_df.append({
            'Year': year,
            'Min': min_value,
            'Max': max_value,
            'Mean': mean_value,
            'StdDev': std_dev
        }, ignore_index=True)

    return stats_df

# Obtain annual summary stats
annual_stats = calculate_annual_statistics_from_monthly_counts(monthly_counts_df)

# %%

def count_unique_neighborhoods(dataframes):
    """
    Count the number of unique neighborhoods for each year.

    Parameters:
    dataframes (dict): A dictionary of DataFrames, keyed by year.

    Returns:
    dict: A dictionary containing the count of unique neighborhoods for each year.
    """
    dataframes = {
        2011: police_df_2011,
        2012: police_df_2012,
        2013: police_df_2013,
        2014: police_df_2014,
        2015: police_df_2015,
        2016: police_df_2016,
        2017: police_df_2017,
        2018: police_df_2018,
        2019: police_df_2019,
        2020: police_df_2020,
        2021: police_df_2021,
        2022: police_df_2022,
    }
    
    unique_counts = {}

    for year, df in dataframes.items():
        unique_neighborhoods = df['Falls within'].nunique()  # Count unique values in 'Falls within'
        unique_counts[year] = unique_neighborhoods

    return unique_counts

# Get the count of unique neighborhoods for each year
unique_neighborhoods_counts = count_unique_neighborhoods(dataframes)

# %%

def find_max_violence_neighborhood(years):
    """
    Find the neighborhood with the highest number of 'Violence and sexual offences' for each year.

    Parameters:
    years (list): A list of years to process.

    Returns:
    dict: A dictionary containing the neighborhood with the maximum count of 'Violence and sexual offences' for each year.
    """
    max_violence_neighborhood = {}

    for year in years:
        # Dynamically access the DataFrame for each year
        df = globals().get(f'police_df_{year}')
        if df is not None:
            # Filter for 'Violence and sexual offences'
            violence_df = df[df['Crime type'] == 'Violence and sexual offences']

            # Group by 'Falls within' and count occurrences, then find the neighborhood with the maximum count
            if not violence_df.empty:
                max_neighborhood = violence_df.groupby('Falls within').size().idxmax()
                max_violence_neighborhood[year] = max_neighborhood
            else:
                max_violence_neighborhood[year] = "No Data"

    return max_violence_neighborhood

# Example usage
# Specify the range of years you have data for
years = range(2011, 2023)

# Find the neighborhood with the highest count of 'Violence and sexual offences' for each year
max_violence_neighborhoods = find_max_violence_neighborhood(years)


# %%

# Filter the DataFrame for the years 2013 to 2022
filtered_df = monthly_counts_df[(monthly_counts_df['Year'] >= 2013) & (monthly_counts_df['Year'] <= 2022)]

# Set the 'Year' column as the index for easier plotting
filtered_df = filtered_df.set_index('Year')

# Plotting
plt.figure(figsize=(10, 6))
for year in filtered_df.index:
    plt.plot(filtered_df.columns, filtered_df.loc[year], label=year)

plt.xlabel('Month', fontweight='bold')
plt.ylabel('Count of Violence and Sexual Offences', fontweight='bold')
plt.title('Monthly Counts of Violence and Sexual Offences from 2013 to 2022', fontweight='bold')
plt.xticks(rotation=45)
plt.legend(title='Year')
plt.grid(True)

filename = 'monthly_counts.png'
save_path = os.path.join(base_path, filename)
plt.savefig(save_path)

plt.show()

# %%

# Set 'Year' as index
annual_stats.set_index('Year', inplace=True)

###
# BAR Plotting
annual_stats.plot(kind='bar', figsize=(12, 6))

plt.title('Annual Crime Statistics', fontweight='bold')
plt.xlabel('Year', fontweight='bold')
plt.ylabel('Statistic Value', fontweight='bold')
plt.xticks(rotation=45)
plt.grid(True)

filename = 'annual_stats_bar.png'
save_path = os.path.join(base_path, filename)
plt.savefig(save_path)

plt.show()


# LINE Plotting

plt.figure(figsize=(12, 6))

plt.plot(annual_stats.index, annual_stats['Min'], label='Min')
plt.plot(annual_stats.index, annual_stats['Max'], label='Max')
plt.plot(annual_stats.index, annual_stats['Mean'], label='Mean')
plt.plot(annual_stats.index, annual_stats['StdDev'], label='StdDev')

# Set x-axis ticks to be each year
plt.xticks(annual_stats.index)

plt.title('Annual Crime Statistics Trends (2013-2022)', fontweight='bold')
plt.xlabel('Year', fontweight='bold')
plt.ylabel('Value', fontweight='bold')
plt.legend(title='Statistic')
plt.grid(True)

filename = 'annual_stats_line.png'
save_path = os.path.join(base_path, filename)
plt.savefig(save_path)

plt.show()

 
    
    

File: Annual_Police_Data.R

# Clear working memory
rm(list=ls())

library(dplyr)
library(readr)
library(ggplot2)
library(reshape2)
library(fs)
library(lubridate)
library(tidyr)

# Set base path and root directory
base_path <- "C:/Users/natal/Desktop/GitHub/SARC_research"
root_directory <- file.path(base_path, "Police_Data")

## Function to loop through the folders and create dataframes according to year
create_police_df_for_year <- function(year, root_directory) {
  # Initialize the DataFrame with specified columns
  police_df <- data.frame(Month=character(), Falls.within=character(), Crime.type=character(), stringsAsFactors=FALSE)
  
  # Iterate through all items in the root directory
  items <- list.files(root_directory, full.names = TRUE)
  for (item in items) {
    # Check if the item is a directory and contains the specified year
    if (file.info(item)$isdir && grepl(year, item)) {
      files <- list.files(item, pattern = "-street\\.csv$", full.names = TRUE)
      for (file in files) {
        # Read the CSV file
        df <- read_csv(file, col_types = cols())
        
        # Check for parsing problems
        parse_problems <- problems(df)
        if (nrow(parse_problems) > 0) {
          print(paste("Parsing issues found in file:", file))
          print(parse_problems)
        }
        
        # Select and rename columns as necessary to match the expected structure
        df <- df %>% 
          select(Month, Falls.within = `Falls within`, Crime.type = `Crime type`)
        
        # Append it to the police_df DataFrame
        police_df <- bind_rows(police_df, df)
      }
    }
  }
  
  return(police_df)
}

# Dataframes from 2011 to 2022
police_df_2011 <- create_police_df_for_year(2011, root_directory)
police_df_2012 <- create_police_df_for_year(2012, root_directory)
police_df_2013 <- create_police_df_for_year(2013, root_directory)
police_df_2014 <- create_police_df_for_year(2014, root_directory)
police_df_2015 <- create_police_df_for_year(2015, root_directory)
police_df_2016 <- create_police_df_for_year(2016, root_directory)
police_df_2017 <- create_police_df_for_year(2017, root_directory)
police_df_2018 <- create_police_df_for_year(2018, root_directory)
police_df_2019 <- create_police_df_for_year(2019, root_directory)
police_df_2020 <- create_police_df_for_year(2020, root_directory)
police_df_2021 <- create_police_df_for_year(2021, root_directory)
police_df_2022 <- create_police_df_for_year(2022, root_directory)


## Function for data exploration
print_unique_values <- function(year, dataframes) {
  # Access the DataFrame for the specified year
  df <- dataframes[[as.character(year)]]
  if (!is.null(df)) {
    # Get and print unique values for "Crime type"
    unique_crime_types <- unique(df$`Crime.type`)
    cat("Unique Crime Types for", year, ":\n")
    print(unique_crime_types)
    
    # Get and print unique values for "Falls within"
    unique_police_neighborhoods <- unique(df$`Falls.within`)
    cat("\nUnique Police Neighborhoods for", year, ":\n")
    print(unique_police_neighborhoods)
  } else {
    cat("No data available for the year", year, "\n")
  }
}


police_df_list <- list(
  "2011" = police_df_2011,
  "2012" = police_df_2012,
  "2013" = police_df_2013,
  "2014" = police_df_2014,
  "2015" = police_df_2015,
  "2016" = police_df_2016,
  "2017" = police_df_2017,
  "2018" = police_df_2018,
  "2019" = police_df_2019,
  "2020" = police_df_2020,
  "2021" = police_df_2021,
  "2022" = police_df_2022
)

# Unique values for each year
print_unique_values(2012, police_df_list)
print_unique_values(2013, police_df_list)
print_unique_values(2014, police_df_list)
print_unique_values(2015, police_df_list)
print_unique_values(2016, police_df_list)
print_unique_values(2017, police_df_list)
print_unique_values(2018, police_df_list)
print_unique_values(2019, police_df_list)
print_unique_values(2020, police_df_list)
print_unique_values(2021, police_df_list)
print_unique_values(2022, police_df_list)

## POPULATION 

# Define the police forces and their populations
police_force <- c("Avon and Somerset Constabulary", "Bedfordshire Police", "British Transport Police",
                  "Cambridgeshire Constabulary", "Cheshire Constabulary", "City of London Police",
                  "Cleveland Police", "Cumbria Constabulary", "Derbyshire Constabulary", "Devon & Cornwall Police",
                  "Dorset Police", "Durham Constabulary", "Dyfed-Powys Police", "Essex Police",
                  "Gloucestershire Constabulary", "Greater Manchester Police", "Gwent Police", "Hampshire Constabulary",
                  "Hertfordshire Constabulary", "Humberside Police", "Kent Police", "Lancashire Constabulary",
                  "Leicestershire Police", "Lincolnshire Police", "Merseyside Police", "Metropolitan Police Service",
                  "Norfolk Constabulary", "North Wales Police", "North Yorkshire Police", "Northamptonshire Police",
                  "Northumbria Police", "Nottinghamshire Police", "Police Service of Northern Ireland",
                  "South Wales Police", "South Yorkshire Police", "Staffordshire Police", "Suffolk Constabulary",
                  "Surrey Police", "Sussex Police", "Thames Valley Police", "Warwickshire Police",
                  "West Mercia Police", "West Midlands Police", "West Yorkshire Police", "Wiltshire Police")

population_served <- c(1650000, 664500, NA, 800000, 1070000, 431000, 570000, 498000, 1060000, 1760000,
                       777000, 641000, 523000, 1860000, 641000, 2850000, 598000, 2000000, 1200000, 934000,
                       1870000, 1520000, 1110000, 766000, 1430000, 8990000, 914000, 703000, 832000, 757000,
                       1470000, 1170000, NA, 1350000, 1420000, 1140000, 761000, 1200000, 1720000, 2430000,
                       584000, 1300000, 2940000, 2350000, 727000)

# Create the data frame
police_region_pop_df <- data.frame(police_force = police_force, population_served = population_served)

# Calculate the total population (excluding NA values)
total_population <- sum(police_region_pop_df$population_served, na.rm = TRUE)

# Add a new row for the total
total_row <- data.frame(police_force = "Total", population_served = total_population)

# Combine the original data frame with the new row
police_region_pop_df <- rbind(police_region_pop_df, total_row)


## Monthly counts and rates of sexual violence

# Function to process a year's data
process_year <- function(df, year, population_served) {
  # Filter for 'Violence and sexual offences'
  filtered_df <- filter(df, `Crime.type` == 'Violence and sexual offences')
  
  # Extract the month from the 'Month' column
  filtered_df$Month <- factor(month(ymd(paste0(filtered_df$Month, "-01")), label = TRUE, abbr = FALSE))
  
  # Count occurrences for each month
  monthly_counts <- filtered_df %>%
    group_by(Month) %>%
    summarise(Count = n(), .groups = 'drop') %>%
    arrange(Month)
  
  # Initialize all months with zero count
  all_months <- setNames(rep(0, 12), month.name)
  counts_list <- as.list(setNames(monthly_counts$Count, monthly_counts$Month))
  new_row <- replace(all_months, names(counts_list), counts_list)
  
  # Add year and calculate crime rates
  new_row_df <- data.frame(Year = year, new_row, stringsAsFactors = FALSE)
  for (month_name in names(all_months)) {
    rate_column <- paste0(month_name, " Crime Rate")
    crime_count <- new_row_df[[month_name]]
    if (!is.na(population_served) && population_served != 0) {
      new_row_df[[rate_column]] <- (crime_count / population_served) * 100000
    } else {
      new_row_df[[rate_column]] <- NA
    }
  }
  
  return(new_row_df)
}

# Get the population served from the police_region_pop_df
population_served <- police_region_pop_df$population_served[46] 

# Adjust the data frame structure to include 'Total Counts' and 'Total Crime Rate'
monthly_rates_df <- data.frame(Year = integer(), January = integer(), February = integer(), 
                               March = integer(), April = integer(), May = integer(), 
                               June = integer(), July = integer(), August = integer(), 
                               September = integer(), October = integer(), November = integer(), 
                               December = integer(), TotalCounts = integer(), TotalCrimeRate = double(), 
                               stringsAsFactors = FALSE)

# Loop to process each year
for (year in 2014:2022) {
  # Retrieve the data frame for the year from the list
  if (!is.null(police_df_list[[as.character(year)]])) {
    yearly_df <- police_df_list[[as.character(year)]]
    
    # Process the data for the year using the 'process_year' function
    new_row <- process_year(yearly_df, year, population_served)
    
    # Calculate the total counts for the year
    total_counts <- sum(new_row[2:13], na.rm = TRUE) # Sum of crimes
    new_row$TotalCounts <- total_counts
    
    # Calculate the total crime rate for the year
    new_row$TotalCrimeRate <- (total_counts / population_served) * 100000 # Crime rate per 100,000 people
    
    # Append the new row to the monthly_rates_df data frame
    monthly_rates_df <- rbind(monthly_rates_df, new_row)
  }
}

# Display the updated data frame
print(monthly_rates_df)


####

# Filtering for 'Violence and sexual offences' in the 2022 data
violence_df_2022 <- police_df_2022 %>%
  filter(`Crime.type` == 'Violence and sexual offences')

# Summing the number of crimes per police force
crime_counts_2022 <- violence_df_2022 %>%
  group_by(`Falls.within`) %>%
  summarise(Total_Crimes = n(), .groups = 'drop')

# Renaming 'Falls.within' to 'police_force' for merging
colnames(crime_counts_2022)[which(names(crime_counts_2022) == "Falls.within")] <- "police_force"

# Merging with the population data
crime_rates_2022 <- merge(crime_counts_2022, police_region_pop_df, by = "police_force")

# Calculating the crime rate per 100,000 population
crime_rates_2022 <- crime_rates_2022 %>%
  mutate(Crime_Rate_per_100k = (Total_Crimes / population_served) * 100000)

# Select only the relevant columns to display
crime_rates_2022 <- crime_rates_2022 %>%
  select(police_force, Total_Crimes, Crime_Rate_per_100k)

# Display the results
print(crime_rates_2022)
